File: src/main/kotlin/com/lambda/client/module/modules/PacketEvent.java

package com.lambda.client.module.modules;

import net.minecraft.network.Packet;
import net.minecraftforge.fml.common.eventhandler.Cancelable;

public class PacketEvent
extends EventStage {
    private int stage;

    public void EventStage() {
    }

    public void EventStage(int stage) {
        this.stage = stage;
    }

    public int getStage() {
        return this.stage;
    }

    public void setStage(int stage) {
        this.stage = stage;
    }
    private final Packet<?> packet;

    public PacketEvent(int stage, Packet<?> packet) {
        super(stage);
        this.packet = packet;
    }

    public <T extends Packet<?>> T getPacket() {
        return (T)this.packet;
    }

    @Cancelable
    public static class Send
    extends PacketEvent {
        public Send(int stage, Packet<?> packet) {
            super(stage, packet);
        }
    }

    @Cancelable
    public static class Receive
    extends PacketEvent {
        public Receive(int stage, Packet<?> packet) {
            super(stage, packet);
        }
    }
}

File: src/main/kotlin/com/defrag/client/util/PacketEvent.kt
package com.defrag.client.util
import com.defrag.client.module.modules.EventStage
import net.minecraft.network.Packet
import net.minecraftforge.fml.common.eventhandler.Cancelable

open class PacketEvent(stage: Int, private val packet: Packet<*>) : EventStage(stage) {

    @Cancelable
    class Send(stage: Int, packet: Packet<*>) : PacketEvent(stage, packet)

    @Cancelable
    class Receive(stage: Int, packet: Packet<*>) : PacketEvent(stage, packet)



    override fun getStage(): Int {
        return stage
    }
    open fun getPacket(): Packet<*> {
        return packet
    }
}
